### PR TITLE
requirement, test: Remove preresolved dependency optimization

### LINKED ISSUE #539: Experiment with removing preresolved dependency optimization
Currently we avoid invoking dependency resolution with hashed dependencies or when `--no-deps` are provided. This causes some inconsistency since `pip` is more accurate at finding dependencies than our hand-coded logic. Now that we've removed our own dependency resolution logic and are just shelling out to `pip`, we should experiment with removing this special case to see what the performance impact is like for large requirements files. If the slowdown is acceptable, we should consider removing the special case and repurposing the hashing logic and `--no-deps` flag to just pass this same flag down to the `pip` invocation.

### LINKED ISSUE #564: Ignore duplicate requirements
## Bug description
With using [pip-tools workflow for layered requirements](https://github.com/jazzband/pip-tools#workflow-for-layered-requirements) pip-audit fails because of duplicate requirements.
## Reproduction steps
```shell
echo "django" > a.in
pip-compile -q --allow-unsafe --generate-hashes --resolver=backtracking --strip-extras a.in
echo -e "-c a.txt\ndjango-debug-toolbar" > b.in
pip-compile -q --allow-unsafe --generate-hashes --resolver=backtracking --strip-extras b.in
pip-audit -r a.txt -r b.txt --require-hashes
```
## Expected behavior
No errors
## Screenshots and logs
```
DEBUG:pip_audit._cli:parsed arguments: Namespace(local=False, requirements=[<_io.TextIOWrapper name='a.txt' mode='r' encoding='UTF-8'>, <_io.TextIOWrapper name='b.txt' mode='r' encoding='UTF-8'>], project_path=None, format=<OutputFormatChoice.Columns: 'columns'>, vulnerability_service=<VulnerabilityServiceChoice.Pypi: 'pypi'>, dry_run=False, strict=False, desc=<VulnerabilityDescriptionChoice.Auto: 'auto'>, cache_dir=None, progress_spinner=<ProgressSpinnerChoice.On: 'on'>, timeout=15, paths=[], verbose=1, fix=False, require_hashes=True, index_url='https://pypi.org/simple/', extra_index_urls=[], skip_editable=False, no_deps=False, output=PosixPath('stdout'), ignore_vulns=[])
DEBUG:pip_audit._cli:Auditing asgiref (3.6.0)
DEBUG:pip_audit._cli:Auditing django (4.1.7)
DEBUG:pip_audit._cli:Auditing sqlparse (0.4.3)
ERROR:pip_audit._cli:package asgiref has duplicate requirements: asgiref==3.6.0 (from RequirementLine(line_number=7, line='asgiref==3.6.0     --hash=sha256:71e68008da809b957b7ee4b43dbccff33d1b23519fb8344e33f049897077afac     --hash=sha256:9567dfe7bd8d3c8c892227827c41cce860b368104c3431da67a0c5a65a949506', filename=PosixPath('/tmp/tmpifyxbix9')))
```
## Platform information

* OS name and version: Arch Linux
* `pip-audit` version (`pip-audit -V`): pip-audit 2.5.2
* Python version (`python -V` or `python3 -V`): Python 3.10.10
* `pip` version (`pip -V` or `pip3 -V`): pip 23.0.1

### PR DESCRIPTION
**Pull request by tetsuo-cpp**
Closes #539.

Closes #564.

### CODE REVIEW

File: test/dependency_source/test_requirement.py [hunk 1]
```diff
@@ -378,109 +427,30 @@ def test_requirement_source_require_hashes_unpinned(req_file):
...
-        SkippedDependency(
-            name="-e file:flask.py",
-            skip_reason="could not deduce package version from URL requirement",
-        )
-        in specs
-    )
-
-
-def test_requirement_source_no_deps_non_editable_without_egg_fragment(req_file):
+def test_requirement_source_require_hashes_incorrect_hash(req_file):
```
Comment (tetsuo-cpp) [inline test/dependency_source/test_requirement.py:430]: I added a few tests but to be honest, some of these (and the existing ones) are essentially testing `pip`. We could potentially trim these down to a bare minimum that at least confirms that we're providing the right flags to `pip` and call it a day.

**Comment by tetsuo-cpp** [comment]
I'm still looking for some large requirements files to test performance with.

File: test/dependency_source/test_requirement.py (hunk 1 above)
Comment (woodruffw) [reply test/dependency_source/test_requirement.py:430]: Yeah, I'd be okay with dropping tests that aren't contributing to our coverage. But yes, it'd be good to confirm that we pass the right flags in some new tests.

File: CHANGELOG.md [hunk 2]
```diff
@@ -15,7 +15,8 @@ All versions prior to 0.0.9 are untracked.
   ([#507](https://github.com/pypa/pip-audit/pull/507))
 
 * `pip-audit`'s handling of dependency resolution has been significantly
-  refactored and simplified ([#523](https://github.com/pypa/pip-audit/pull/523))
+  refactored and simplified ([#523](https://github.com/pypa/pip-audit/pull/523)
+  and [#540](https://github.com/pypa/pip-audit/pull/540))
```
Comment (woodruffw) [inline CHANGELOG.md:19]: NB: We'll probably want a separate `CHANGELOG` entry after all, since we're making this change after this entry's release.

**Review by woodruffw — APPROVED** [review]
LGTM! Let's get this deconflicted and merged 🙂

File: CHANGELOG.md [hunk 3]
```diff
@@ -31,7 +31,8 @@ All versions prior to 0.0.9 are untracked.
   ([#507](https://github.com/pypa/pip-audit/pull/507))
 
 * `pip-audit`'s handling of dependency resolution has been significantly
-  refactored and simplified ([#523](https://github.com/pypa/pip-audit/pull/523))
+  refactored and simplified ([#523](https://github.com/pypa/pip-audit/pull/523)
```
Comment (woodruffw) [inline CHANGELOG.md:34]: Let's break this into a separate CHANGELOG entry, now that the previous entry is part of a past release.

Comment (tetsuo-cpp) [reply CHANGELOG.md]: [05180d3](https://github.com/pypa/pip-audit/pull/540/commits/05180d3f9f36e643958571bd076104f3a252a798)

**Comment by tetsuo-cpp** [comment]
I didn't manage to find any ridiculously large requirements files. But I tested across a few and found that there isn't a meaningful performance impact vs just a regular `pip install` invocation.

**Comment by woodruffw** [comment]
Glad to hear it! Yeah, I think we're fine to move ahead here -- feel free to do another patch release to get this out 🙂

**Pull request merged**